### PR TITLE
Radio button, radio group components

### PR DESCRIPTION
**Pull request by kirill-kolombet**
Finished the radio button component (ts-radio, ts-radio-group). I would like to get any feedback, how to make it better. 

Currently, it supports all features of _ts-ui_ radio button component and has the same visual style. It support controlling via keyboard arrows when element focused. The internal structure of _ts-radio-group_ described as ts-radio HTML elements in the slot.

I have one idea: maybe we can make it more declarative and describe a structure with the data property. It's not so useful feature, but we can add it later.
`<ts-radio-group data="[{
	label,
	value,
	checked,
	disabled
}, ...]"> </ts-radio-group>`

### CODE REVIEW

File: packages/components/radio-group/src/radio-group.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+			if (enabled.length === 0) {
+				return;
+			}
+
+			const direction = getDirectionByKey(e.code);
+			if (!direction) {
+				return;
+			}
+
+			const getNextCheckbox = direction => {
```
Comment (tradeshiftci) [inline packages/components/radio-group/src/radio-group.js]: MAJOR "direction" hides or potentially hides a variable declared in an outer scope at line 91. [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3AVariableShadowing)

File: packages/components/radio-group/src/radio-group.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+				this.index = this.index + direction;
+				if (this.index <= -1) this.index = this.radios.length - 1;
```
Comment (tradeshiftci) [inline packages/components/radio-group/src/radio-group.js]: MAJOR Add curly braces around the nested statement(s) in this "if" block. [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3ACurlyBraces)

File: packages/components/radio-group/src/radio-group.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+				if (this.index >= this.radios.length) this.index = 0;
```
Comment (tradeshiftci) [inline packages/components/radio-group/src/radio-group.js]: MAJOR Add curly braces around the nested statement(s) in this "if" block. [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3ACurlyBraces)

File: packages/components/radio-group/src/radio-group.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+				this.toggle(this.radios[this.index]);
+			}
+		}
+
+		handleKeyDown(e) {
+			if (this.radios.length === 0) {
+				return;
+			}
+
+			const enabled = this.radios.filter(checkbox => !checkbox.disabled);
```
Comment (tradeshiftci) [inline packages/components/radio-group/src/radio-group.js]: MAJOR "checkbox" hides or potentially hides a variable declared in an outer scope at line 102. [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3AVariableShadowing)

File: packages/components/radio-group/src/utils/helper-functions.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import { keyboardKeys } from './constants';
+
+export default function getDirectionByKey(keyCode) {
+	switch (keyCode) {
```
Comment (tradeshiftci) [inline packages/components/radio-group/src/utils/helper-functions.js]: MAJOR Add a "default" clause to this "switch" statement. [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3ASwitchWithoutDefault)

File: index.html [hunk 6]
```diff
@@ -65,686 +68,33 @@
...
+			<form action="/index.php">
+				<!--				<fieldset class="radio-button-container" role="radiogroup" aria-labelledby="group_label_1">-->
+				<ts-radio-group name="radio">
+					<ts-radio value="blue" label="Blue pill"></ts-radio>
+					<ts-radio value="yellow" label="Yellow pill"></ts-radio>
+					<ts-radio value="black" label="Black pill" disabled></ts-radio>
+					<ts-radio value="white" label="White pill" checked></ts-radio>
+					<ts-radio value="green" label="Green pill"></ts-radio>
+				</ts-radio-group>
+				<!--				</fieldset>-->
```
Comment (tradeshiftci) [inline index.html]: MAJOR Remove this commented out code. [rule](https://sonar.ts.sv/coding_rules#rule_key=Web%3AAvoidCommentedOutCodeCheck)
MINOR Add a space at column 23. [rule](https://sonar.ts.sv/coding_rules#rule_key=Web%3AWhiteSpaceAroundCheck)

File: index.html [hunk 7]
```diff
@@ -65,686 +68,33 @@
...
-			}
-			document.getElementById('my-aside').addEventListener('open', onOpen);
-			document.getElementById('my-aside').addEventListener('opened', onOpened);
-			document.getElementById('my-aside').addEventListener('close', onClose);
-			document.getElementById('my-aside').addEventListener('closed', onClosed);
-		</script>
+		<article>
+			<h3>Radio</h3>
+			<form action="/index.php">
+				<!--				<fieldset class="radio-button-container" role="radiogroup" aria-labelledby="group_label_1">-->
```
Comment (tradeshiftci) [inline index.html]: MAJOR Remove this commented out code. [rule](https://sonar.ts.sv/coding_rules#rule_key=Web%3AAvoidCommentedOutCodeCheck)
MINOR Add a space at column 103. [rule](https://sonar.ts.sv/coding_rules#rule_key=Web%3AWhiteSpaceAroundCheck)

File: packages/components/radio/src/utils/index.js [hunk 8]
```diff
@@ -0,0 +1,3 @@
+import * as constantsData from './constants';
```
Comment (tradeshiftci) [inline packages/components/radio/src/utils/index.js]: MAJOR Explicitly import the specific member needed. [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3AS2208)

**Comment by tradeshiftci** [comment]
SonarQube analysis reported 11 issues
* ![MAJOR][MAJOR] 8 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [helper-functions.js#](https://github.com/Tradeshift/elements/blob/dd90da10921ec72d8377499c85088b8cc61b9aaa/packages/components/radio-group/src/utils/helper-functions.js#): Rename this file to "getDirectionByKey". [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3AS3317)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: packages/components/radio-group/src/radio-group.js [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+				this.toggle(this.radios[this.index]);
+			}
+		}
+
+		handleKeyDown(e) {
+			if (this.radios.length === 0) {
+				return;
+			}
+
+			const enabledRadios = this.radios.filter(checkbox => !checkbox.disabled);
```
Comment (tradeshiftci) [inline packages/components/radio-group/src/radio-group.js]: MAJOR "checkbox" hides or potentially hides a variable declared in an outer scope at line 102. [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3AVariableShadowing)

**Comment by tradeshiftci** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [helper-functions.js#](https://github.com/Tradeshift/elements/blob/cdbf3578e5bcd1ffcdd449b44bdaa90fe2c335ae/packages/components/radio-group/src/utils/helper-functions.js#): Rename this file to "getDirectionByKey". [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3AS3317)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by tradeshiftci** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [index.html#](https://github.com/Tradeshift/elements/blob/f0fe779c367c8cd3c58da539347c47715b3356c5/index.html#): Current file has 1007 lines, which is greater than 1000 authorized. Split it into smaller files. [rule](https://sonar.ts.sv/coding_rules#rule_key=Web%3AFileLengthCheck)
1. ![MINOR][MINOR] [helper-functions.js#](https://github.com/Tradeshift/elements/blob/f0fe779c367c8cd3c58da539347c47715b3356c5/packages/components/radio-group/src/utils/helper-functions.js#): Rename this file to "getDirectionByKey". [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3AS3317)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by tradeshiftci** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [index.html#](https://github.com/Tradeshift/elements/blob/cbef3182fa2367e36bb23b6c35f18e6771c90f70/index.html#): Current file has 1002 lines, which is greater than 1000 authorized. Split it into smaller files. [rule](https://sonar.ts.sv/coding_rules#rule_key=Web%3AFileLengthCheck)
1. ![MINOR][MINOR] [helper-functions.js#](https://github.com/Tradeshift/elements/blob/cbef3182fa2367e36bb23b6c35f18e6771c90f70/packages/components/radio-group/src/utils/helper-functions.js#): Rename this file to "getDirectionByKey". [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3AS3317)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by tradeshiftci** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [index.html#](https://github.com/Tradeshift/elements/blob/b3dbc9a0bea79b28d6a51ebcc9dfcdfe80a4c09f/index.html#): Current file has 1002 lines, which is greater than 1000 authorized. Split it into smaller files. [rule](https://sonar.ts.sv/coding_rules#rule_key=Web%3AFileLengthCheck)
1. ![MINOR][MINOR] [helper-functions.js#](https://github.com/Tradeshift/elements/blob/b3dbc9a0bea79b28d6a51ebcc9dfcdfe80a4c09f/packages/components/radio-group/src/utils/helper-functions.js#): Rename this file to "getDirectionByKey". [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3AS3317)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: packages/components/radio-group/src/radio-group.css [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+		border: 0 none;
+	}
+
+	& > fieldset > div {
+		padding: var(--ts-unit-quarter);
+		border-radius: 3px;
+	}
+}
+
+:host(:not([focused])) {
```
Comment (esmailpour-hosein) [inline packages/components/radio-group/src/radio-group.css]: What's the reason for this specific rule?

File: packages/components/radio/src/radio.css [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+:host {
+	& label {
+		display: block;
+		position: relative;
+		padding-left: var(--ts-unit-plus);
+		margin: var(--ts-unit-half);
+		cursor: pointer;
+		font-size: var(--ts-fontsize);
+		-webkit-user-select: none;
```
Comment (esmailpour-hosein) [inline packages/components/radio/src/radio.css]: We have postcss, no need for browser prefixes.

File: packages/components/radio/src/radio.css [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+		& input {
+			position: absolute;
+			opacity: 0;
+			cursor: pointer;
+			height: 0;
+			width: 0;
+		}
+
+		& span {
```
Comment (esmailpour-hosein) [inline packages/components/radio/src/radio.css]: Please use class names, it make it easier to understand the styles. 
This is a topic that we need to discuss later in team and add it to the guideline.

File: packages/components/radio/src/radio.css [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+/* Disable status .................................................*/
+:host([disabled]) {
+	color: var(--ts-disabled-background-color);
+	cursor: not-allowed;
+
+	&:hover {
+		& span {
+			background-color: transparent;
```
Comment (esmailpour-hosein) [inline packages/components/radio/src/radio.css]: It seems that the disabled state of the radio still have the hover effect.

File: packages/components/radio/src/radio.js [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+			return [TSElement.styles, unsafeCSS(css)];
+		}
+
+		static get properties() {
+			return {
+				value: { type: String, reflect: true },
+				label: { type: String, reflect: true },
+				checked: { type: Boolean, reflect: true },
+				disabled: { type: Boolean, reflect: true },
+				dir: { type: String, reflect: true }
```
Comment (esmailpour-hosein) [inline packages/components/radio/src/radio.js]: Why did you add the dir to radio itself? I think only radio-group should have the dir.

File: packages/components/radio/src/radio.js [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+			};
+		}
+
+		handleClick(e) {
+			e.preventDefault();
+			e.stopPropagation();
+			this.dispatchEvent(new Event(customEventNames.RADIO_CLICK, { bubbles: true, composed: true }));
```
Comment (esmailpour-hosein) [inline packages/components/radio/src/radio.js]: Why not using the `this.dispatchCustomEvent`?

File: packages/components/radio/src/radio.js [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+		}
+
+		get direction() {
+			return this.dir || this.bodyDir;
+		}
+
+		handleFocus() {
+			this.dispatchEvent(new Event(customEventNames.RADIO_FOCUS, { bubbles: true, composed: true }));
```
Comment (esmailpour-hosein) [inline packages/components/radio/src/radio.js]: this.dispatchCustomEvent

File: packages/components/radio/src/radio.js [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+		}
+
+		handleBlur() {
+			this.dispatchEvent(new Event(customEventNames.RADIO_BLUR, { bubbles: true, composed: true }));
```
Comment (esmailpour-hosein) [inline packages/components/radio/src/radio.js]: this.dispatchCustomEvent

File: packages/components/radio/src/radio.js [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+					${this.label}
+					<input
+						tabindex="-1"
+						type="radio"
+						role="radio"
+						.value="${this.value}"
+						?checked="${this.checked}"
+						?disabled="${this.disabled}"
+					/>
+					<span></span>
```
Comment (esmailpour-hosein) [inline packages/components/radio/src/radio.js]: Can we use pseudo elements instead? `:after`?

File: packages/core/src/vars.css [hunk 19]
```diff
@@ -169,6 +169,12 @@
 	--spinner-color-blue: var(--ts-color-blue);
 	--spinner-color-mono: var(--ts-color-slate-lightest);
 	--spinner-color-white: var(--ts-color-white);
+
+	/* Radio button........................................................ */
+
+	--radio-border-selected: 7px;
```
Comment (esmailpour-hosein) [inline packages/core/src/vars.css]: You should use one of the variables. I think ts-unit-quarter.

File: packages/core/src/vars.css [hunk 20]
```diff
@@ -169,6 +169,12 @@
...
+	--ts-radio-boxshadow-border-width: inset 0 0 0 var(--radio-border-selected);
+	--ts-radio-unselected-boxshadow-border-width: inset 0 0 0 1.5px;
```
Comment (esmailpour-hosein) [inline packages/core/src/vars.css]: why 1.5px?

File: packages/components/radio/src/radio.css [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+			width: var(--ts-unit);
+			border-radius: 50%;
+
+			&:after {
+				content: '';
+				position: absolute;
+				display: none;
+				top: var(--ts-unit-quarter);
+				left: var(--ts-unit-quarter);
+				width: 8px;
```
Comment (esmailpour-hosein) [inline packages/components/radio/src/radio.css]: Please use variable and maybe some division of unit

File: packages/components/radio-group/src/utils/helper-functions.js [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+		case keyboardKeys.ARROW_RIGHT:
+		case keyboardKeys.ARROW_DOWN:
+			return 1;
```
Comment (esmailpour-hosein) [inline packages/components/radio-group/src/utils/helper-functions.js]: These codes are not easy to understand, can you please change them to some constants that shows what they mean?

File: packages/components/radio/src/radio.js (hunk 14 above)
Comment (kirill-kolombet) [reply packages/components/radio/src/radio.js]: Yes, it's much more logical to set dir only on the parent. The current implementation hasn't changed from the first version, where each radio was an independent entity.

Comment (kirill-kolombet) [reply packages/components/radio/src/radio.js]: Fixed

File: packages/components/radio/src/radio.css (hunk 13 above)
Comment (kirill-kolombet) [reply packages/components/radio/src/radio.css]: The current design of _ts-ui_ doesn't have any hover states, so I removing them here too.

Comment (kirill-kolombet) [reply packages/components/radio/src/radio.css]: Done

File: packages/components/radio/src/radio.css (hunk 12 above)
Comment (kirill-kolombet) [reply packages/components/radio/src/radio.css]: Sure. I thought we prefer to use attributes, properties, and structure over class names (for example button element).  But It could be better to use move descriptive class names where it is possible.

Do we have any special naming rules for classes? I suppose I would call it "checkbox" or "flag".

File: packages/components/radio/src/radio.css (hunk 11 above)
Comment (kirill-kolombet) [reply packages/components/radio/src/radio.css]: Fixed

File: packages/components/radio-group/src/radio-group.css (hunk 10 above)
Comment (kirill-kolombet) [reply packages/components/radio-group/src/radio-group.css]: No reason, it could be easy replaced with just normal hover state without any selector.
RadioGroup have the blue border shadow on hover in _ts-ui_ design. Could be 

```
:host > fieldset > div:hover {
			box-shadow: var(--ts-boxshadow-border-width-focused) var(--ts-color-blue-light);
		}
```

File: packages/components/radio-group/src/radio-group.js [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+				`;
+			return html`
+				<fieldset
+					role="radiogroup"
+					tabindex="0"
+					@keydown="${this.handleKeyDown}"
+					@focus="${this.handleFocus}"
+					@blur="${this.handleBlur}"
+				>
+					${title}
```
[the remark this answers is not in the document]
Comment (kirill-kolombet) [reply packages/components/radio-group/src/radio-group.js]: Fixed. Also should be at right on RTL and have different font-style in _ts-ui_ design

File: packages/components/radio/src/radio.js (hunk 18 above)
Comment (kirill-kolombet) [reply packages/components/radio/src/radio.js]: Yes, we can. Also, I am not sure if we need to have `<input>` element inside. It has no graphical purpose and placed here only for semantics. I suppose we can reduce radio to a single label element.

File: packages/core/src/vars.css (hunk 19 above)
Comment (kirill-kolombet) [reply packages/core/src/vars.css]: I made it 7px to make it identical to _ts-ui_. If it's not required, surely I'll use ts-unit variable

File: packages/core/src/vars.css (hunk 20 above)
Comment (kirill-kolombet) [reply packages/core/src/vars.css]: Copied those styles from _ts-ui_. We don't have new designs yet, so it's temporary values. Changed them to standard vars values.

File: packages/components/radio/src/radio.css (hunk 21 above)
Comment (kirill-kolombet) [reply packages/components/radio/src/radio.css]: Done

File: packages/components/radio-group/src/utils/helper-functions.js (hunk 22 above)
Comment (kirill-kolombet) [reply packages/components/radio-group/src/utils/helper-functions.js]: I suggest having DIRECTION_FORWARD and DIRECTION_BACKWARD constants. Or just FORWARD and BACKWARD. Could the default zero value be called DIRECTION_NONE?

File: package.json [hunk 24]
```diff
@@ -156,6 +153,11 @@
   "publishConfig": {
     "access": "public"
   },
+  "contributors": [
+    "Daniel Spitzer <dsp@tradeshift.com> (https://github.com/sampi)",
+    "Leo Zhang <lza@tradeshift.com> (https://github.com/zdlm)",
+    "Kian (Hossein) Esmaeilpour <hossein.esmaeil@tradeshift.com> (https://github.com/esmailpour-hosein)"
```
[the remark this answers is not in the document]
Comment (kirill-kolombet) [reply package.json]: Ok, I'll add it in the separate PR later 💯 
And I am sorry for this change, some tool (I suppose IDEA) constantly changing package.json keys order.

File: packages/components/radio-group/src/utils/helper-functions.js (hunk 22 above)
Comment (kirill-kolombet) [reply packages/components/radio-group/src/utils/helper-functions.js]: Done

File: packages/components/radio-group/src/radio-group.css (hunk 10 above)
Comment (kirill-kolombet) [reply packages/components/radio-group/src/radio-group.css]: Fixed

File: packages/components/radio/src/radio.css (hunk 12 above)
Comment (kirill-kolombet) [reply packages/components/radio/src/radio.css]: Fixed

File: packages/components/radio/src/radio.js (hunk 18 above)
Comment (kirill-kolombet) [reply packages/components/radio/src/radio.js]: Fixed

File: packages/components/radio-group/src/utils/helper-functions.js [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+	}
+}
+
+export function getNextElementIndex(direction, index, collectionLength) {
+	switch (direction) {
+		case directionTypes.FORWARD:
+			if (index === collectionLength - 1) {
+				return 0;
+			}
+			return ++index;
```
Comment (tradeshiftci) [inline packages/components/radio-group/src/utils/helper-functions.js]: MAJOR Extract this increment operation into a dedicated statement. [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3AS881)

File: packages/components/radio-group/src/utils/helper-functions.js [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+		case directionTypes.BACKWARD:
+			if (index === 0) {
+				return collectionLength - 1;
+			}
+			return --index;
```
Comment (tradeshiftci) [inline packages/components/radio-group/src/utils/helper-functions.js]: MAJOR Extract this decrement operation into a dedicated statement. [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3AS881)

**Comment by tradeshiftci** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: packages/components/radio/src/radio.js (hunk 17 above)
Comment (kirill-kolombet) [reply packages/components/radio/src/radio.js]: Fixed

File: packages/components/radio/src/radio.js (hunk 16 above)
Comment (kirill-kolombet) [reply packages/components/radio/src/radio.js]: Fixed

File: packages/components/radio/src/radio.js (hunk 15 above)
Comment (kirill-kolombet) [reply packages/components/radio/src/radio.js]: Fixed

File: package.json (hunk 24 above)
[the remark this answers is not in the document]
Comment (esmailpour-hosein) [reply package.json]: NP :) 
Probably this one does not work and we need to check it out :
```    
    "package.json": [
      "npx sort-package-json"
    ]

File: packages/components/radio/src/radio.css (hunk 12 above)
Comment (esmailpour-hosein) [reply packages/components/radio/src/radio.css]: > I thought we prefer to use attributes, properties, and structure over class names (for example button element). But It could be better to use move descriptive class names where it is possible.

Yes, when there's an attribute or property that already have a clear name then that should be enough, but using tag name as selector make it hard to understand what it does!

>Do we have any special naming rules for classes? 
No, at least not yet!

File: packages/components/radio/src/radio.js (hunk 18 above)
Comment (esmailpour-hosein) [reply packages/components/radio/src/radio.js]: >Yes, we can. Also, I am not sure if we need to have <input> element inside. It has no graphical purpose and placed here only for semantics. I am not sure if semantics have any sense inside shadow dom, I suppose we can reduce radio template to a single label element.


Can you please check them in form data? I think we need to keep the inputs anyway for future proofing the input related events!

File: packages/core/src/vars.css (hunk 19 above)
Comment (esmailpour-hosein) [reply packages/core/src/vars.css]: Please ask @mihneaz or Vlad to get a confirmation about the size, but it should be at least 
 similar to ts-ui if not identical 👍

File: packages/core/src/vars.css (hunk 20 above)
Comment (esmailpour-hosein) [reply packages/core/src/vars.css]: Don't let legacy code hold you back ;)

File: packages/components/radio/README.md [hunk 27]
```diff
@@ -0,0 +1,147 @@
...
+
+## ➤ Properties
+
+| Property | Attribute | Type    | Default | Description              |
+| -------- | --------- | ------- | ------- | ------------------------ |
+| value    | value     | string  | ''      | Value of the radio       |
+| label    | data      | string  | ''      | Label of the radio       |
+| checked  | checked   | boolean | false   | Check status             |
+| disabled | disabled  | boolean | false   | Disabled status          |
+| dir      | dir       | string  | ''      | Direction 'rtl' or 'ltr' |
```
Comment (esmailpour-hosein) [inline packages/components/radio/README.md]: do we need to have dir on radio itself too? I think it should be enough to have it on radio-group

File: packages/components/radio/src/radio.css [hunk 28]
```diff
@@ -0,0 +1,53 @@
...
+		box-shadow: var(--ts-radio-boxshadow-border-width) var(--ts-color-gray-lighter);
+	}
+}
+
+/* rtl ............................................................*/
+:host([dir='rtl']) {
+	& label {
+		padding-right: var(--ts-unit-plus);
+
+		& span {
```
Comment (esmailpour-hosein) [inline packages/components/radio/src/radio.css]: can you please change this one too?

File: packages/components/radio/stories/radio.stories.js [hunk 29]
```diff
@@ -0,0 +1,17 @@
...
+
+storiesOf('ts-radio', module)
+	.addDecorator(withKnobs)
+	.add('default', () => {
+		const label1 = text('label1', 'Red pill');
+		const label2 = text('label2', 'Blue pill');
+		const disabled = boolean('Disabled', false);
+
+		return html`
+			<ts-radio ?disabled="${disabled}" label="${label1}"></ts-radio>
```
Comment (esmailpour-hosein) [inline packages/components/radio/stories/radio.stories.js]: please add all available properties and  make it more dynamic

Comment (kirill-kolombet) [reply packages/components/radio/stories/radio.stories.js]: Radio group has properties, like "value" and "index" which meant to be read-only. Added 'title' and 'focused' properties. How can we make more dynamic? Ability to customize the number of radio inputs?

File: packages/components/radio-group/src/radio-group.js [hunk 30]
```diff
@@ -0,0 +1,143 @@
...
+			this.focused = false;
+		}
+
+		handleSlotChange() {
+			this.radios = this.radioElements;
+
+			this.index = 0;
+			for (let i = 0; i < this.radios.length; i++) {
+				const radio = this.radios[i];
+				radio.dir = this.direction;
```
[the remark this answers is not in the document]
Comment (kirill-kolombet) [reply packages/components/radio-group/src/radio-group.js]: If the direction of _ts-radio-group_ changed in runtime, we need to rerender all slotted _ts-radio_ components to apply direction change. Is there any possibility to force rendering of children components without changing their properties?

[the remark this answers is not in the document]
Comment (hossein-esmaeilpourarabi) [reply packages/components/radio-group/src/radio-group.js]: For direction we talked and decided it's okay to not be super responsive to the changes like this since they are really rare 👍 most of the times users are seeing these changes after changing the language and need to re-login to see changes.

**Comment by kirill-kolombet** [comment]
Current design of the ts-radio-group:
Screenshot 2020-01-06 at 13 47 38

File: packages/components/radio/src/radio.js (hunk 18 above)
Comment (esmailpour-hosein) [reply packages/components/radio/src/radio.js]: @kirill-kolombet What was the result?

Comment (kirill-kolombet) [reply packages/components/radio/src/radio.js]: I created a simple form to check the behavior of inputs inside radio shadow DOM. If the input element inside, it's content not accessible with the form data

```
	let form = document.querySelector('#radio-form');
	form.addEventListener('submit', (e) => {
		e.preventDefault();
		const form = document.querySelector('form');
		const data = Object.fromEntries(new FormData(form).entries()); //data is empty
	});
```

Currently, the only way to get the value of the radio group is to subscribe to the event: 
```
    this.refs['radio'].addEventListener(
    	'radio-selected', (e) => { this.setState({value: e.detail.value}); }
    );
```

Or directly get value on submit: 
```
    const value = this.refs['radio'].value;
```

If we want ts-radio-group to work with the form data we can:
1. Use hidden <input> element instead of <ts-radio> inside <ts-radio-group> slot. Generate visual representation of those inputs with the <ts-radio> in the shadow DOM, and update slotted inputs state on change
2. Generate hidden <input> with unique id element in the light DOM, next to <ts-radio-group>, like it's done in the weightless components. The simple and effective way.

Comment (kirill-kolombet) [reply packages/components/radio/src/radio.js]: I would like to delay this decision until we'll make more custom input controls and look at what would work best for us.
